service/frontend/workflowHandler.go

```diff
@@ -1067,38 +1067,75 @@ func (wh *WorkflowHandler) RespondActivityTaskCanceledByID(
 // RespondDecisionTaskCompleted - response to a decision task
 func (wh *WorkflowHandler) RespondDecisionTaskCompleted(
 	ctx context.Context,
-	completeRequest *gen.RespondDecisionTaskCompletedRequest) error {
+	completeRequest *gen.RespondDecisionTaskCompletedRequest) (*gen.RespondDecisionTaskCompletedResponse, error) {
 
 	scope := metrics.FrontendRespondDecisionTaskCompletedScope
 	sw := wh.startRequestProfile(scope)
 	defer sw.Stop()
 
 	if completeRequest == nil {
-		return wh.error(errRequestNotSet, scope)
+		return nil, wh.error(errRequestNotSet, scope)
 	}
 
 	// Count the request in the RPS, but we still accept it even if RPS is exceeded
 	wh.rateLimiter.TryConsume(1)
 
 	if completeRequest.TaskToken == nil {
-		return wh.error(errTaskTokenNotSet, scope)
+		return nil, wh.error(errTaskTokenNotSet, scope)
 	}
 	taskToken, err := wh.tokenSerializer.Deserialize(completeRequest.TaskToken)
 	if err != nil {
-		return wh.error(err, scope)
+		return nil, wh.error(err, scope)
 	}
 	if taskToken.DomainID == "" {
-		return wh.error(errDomainNotSet, scope)
+		return nil, wh.error(errDomainNotSet, scope)
 	}
 
-	err = wh.history.RespondDecisionTaskCompleted(ctx, &h.RespondDecisionTaskCompletedRequest{
+	histResp, err := wh.history.RespondDecisionTaskCompleted(ctx, &h.RespondDecisionTaskCompletedRequest{
 		DomainUUID:      common.StringPtr(taskToken.DomainID),
 		CompleteRequest: completeRequest},
 	)
 	if err != nil {
-		return wh.error(err, scope)
+		return nil, wh.error(err, scope)
 	}
-	return nil
+
+	completedResp := &gen.RespondDecisionTaskCompletedResponse{}
+
+	if completeRequest.GetReturnNewDecisionTask() && histResp != nil && histResp.StartedResponse != nil {
+		startedResponse := histResp.StartedResponse
+		matchingResp := &m.PollForDecisionTaskResponse{}
```

Comment:
Can you put this logic in a helper function, and invoke it from both matchingEngine.PollForDecisionTask and frontend.RespondDecisionTaskCompleted?